### PR TITLE
notion: improvement

### PR DESCRIPTION
**Pull request by tftio**
Builds the notion (not-ion) window manager. Tested with and without pango/xft patch on my local Linux system.

### CODE REVIEW

**Comment by jfb** [comment]
This is my first submission. If there's anything wrong here, I'll happily fix. One issue is the location of the -pango diff; I'm happy to leave it on github, but if there's a better place, I'm happy to use that, too.

File: pkgs/applications/window-managers/notion/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{ pangoSupport ? true, stdenv, fetchurl, x11, lua, gettext, groff, pkgconfig, busybox, libXinerama, libXrandr, libX11, xmessage, fontconfig ? null, patches ? [], fallbackFont ? "Sans" }:
+
+assert pangoSupport -> fontconfig != null;
+
+let
+pangoPatch = fetchurl {
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/default.nix]: Put the patch in the nixpkgs tree please.

**Comment by lethalman** [comment]
Why is it called pangoSupport if I don't see any pango dependency?

**Comment by jfb** [comment]
That's the patch name. In reality, it's adding xft support. I'll rename appropriately.

**Comment by jfb** [comment]
Should I simply commit the patch into nixpkgs/.../notion/<patchname.diff>?

**Comment by lethalman** [comment]
@jfb yes please. You can refer to it like `patches = [ ./foo.patch ]`.

**Comment by jfb** [comment]
Updated with @lethalman feedback.

**Comment by jfb** [comment]
on it.

File: pkgs/applications/window-managers/notion/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    platforms = platforms.linux;
+    license   = licenses.notion_lgpl;
+    maintainers = [maintainers.jfb];
+  };
+  src = fetchurl {
+    url = https://github.com/raboof/notion/archive/3-2015061300.tar.gz;
+    sha256 = "3c9d9f35a9fb0d17c263b76fe28f7a1a4a05b7d6140545524cc1effd98c5c305";
+  };
+
+  patches = patches ++ stdenv.lib.optional enableXft xftPatch;
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/default.nix]: The patch file must NOT be a string, also just put it inline: `patches = patches ++ stdenv.lib.optional enableXft ./notion-xft_nixos.diff;`

File: pkgs/applications/window-managers/notion/default.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  lua, gettext, groff,
+  pkgconfig, busybox,
+  x11, libXinerama, libXrandr, libX11
+}:
+
+assert enableXft -> fontconfig != null && fallbackFont != null;
+
+let
+  xftPatch = "./notion-xft_nixos.diff";
+  name     = "notion";
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/default.nix]: Put name and version directly in the derivation, it's hard to read with a let expression.

File: pkgs/applications/window-managers/notion/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    maintainers = [maintainers.jfb];
+  };
+  src = fetchurl {
+    url = https://github.com/raboof/notion/archive/3-2015061300.tar.gz;
+    sha256 = "3c9d9f35a9fb0d17c263b76fe28f7a1a4a05b7d6140545524cc1effd98c5c305";
+  };
+
+  patches = patches ++ stdenv.lib.optional enableXft xftPatch;
+
+#   postPatch = "
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/default.nix]: Why is this commented? If it's unneeded just remove it.

**Comment by lethalman** [comment]
Some other nitpicks :)

**Comment by jfb** [comment]
You beat me to those nitpicks! I'll pull name and version into the meta instead of the let expr.

File: pkgs/applications/window-managers/notion/default.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    platforms = platforms.linux;
+    license   = licenses.notion_lgpl;
+    maintainers = [maintainers.jfb];
+  };
+  src = fetchurl {
+    url = https://github.com/raboof/notion/archive/3-2015061300.tar.gz;
+    sha256 = "3c9d9f35a9fb0d17c263b76fe28f7a1a4a05b7d6140545524cc1effd98c5c305";
+  };
+
+  patches = patches ++ stdenv.lib.optional enableXft [./notion-xft_nixos.diff];
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/default.nix]: You don't like copy-pasting, do you? :) `lib.optional` already returns a list. Just `patches = patches ++ stdenv.lib.optional enableXft ./notion-xft_nixos.diff;` as suggested before should work.

File: pkgs/applications/window-managers/notion/default.nix [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  postPatch = "substituteInPlace system-autodetect.mk --replace '#PRELOAD_MODULES=1' 'PRELOAD_MODULES=1'";
+  buildInputs = [x11 lua gettext groff pkgconfig busybox libXinerama libXrandr libX11] ++ stdenv.lib.optional enableXft [libXft];
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/default.nix]: Same here, just `libXft`

File: pkgs/applications/window-managers/notion/notion-xft_nixos.diff [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+ ##
++## Xft support
++##
++
++USE_XFT=1
++
++ifeq ($(USE_XFT),1)
++EXTRA_INCLUDES += `pkg-config xft --cflags`
++EXTRA_LIBS += `pkg-config xft --libs`
++DEFINES += -DXFT -DCF_FALLBACK_FONT_NAME=\"${fallbackFont}\"
```
Comment (lethalman) [inline pkgs/applications/window-managers/notion/notion-xft_nixos.diff]: Are you sure this works? You don't pass this `fallbackFont` variable anywhere through the derivation. I think you missed a `inherit fallbackFont` in the drv.

Comment (jfb) [inline pkgs/applications/window-managers/notion/notion-xft_nixos.diff]: In fact, I'm sure it doesn't work. Patch incoming.

File: pkgs/applications/window-managers/notion/default.nix (hunk 5 above)
Comment (jfb) [inline pkgs/applications/window-managers/notion/default.nix]: I'm still learning nix-the-language. Thanks for being so patient.

**Comment by lethalman** [comment]
Is this ready or you still need to do further changes?

**Comment by jfb** [comment]
I believe it's ready. Builds both with and without xft locally. Ship it!

**Comment by lethalman** [comment]
Looks good. Thanks and welcome :)

**Comment by jfb** [comment]
Whoo hoo!

**Pull request merged**